### ISSUE TITLE
jnp.linalg.eigh not supported

### ISSUE THREAD

**Issue by silvasean** [openxla/openxla-pjrt-plugin#25]
## Expected Behavior

The following program does not crash with a compiler error:

```
import jax.numpy as jnp
a = jnp.ones((702, 4, 4), dtype=jnp.float32)
print(a.shape, a.dtype)
print(jnp.linalg.eigh(a))
```

## Actual Behavior

```
jax._src.source_info_util.JaxStackTraceBeforeTransformation: NotImplementedError: MLIR translation rule for primitive 'eigh' not found for platform iree_cpu
```

Full error log:
https://gist.github.com/silvasean/51d41a3685fbf52fa708d75dd37ed078

I tried subbing out the CPU/GPU/TPU paths, but they all fall back on custom calls at some level which we don't support.

CPU/GPU paths need a `syevd` blas implementation [link](https://github.com/google/jax/blob/f461c4ef0c0c73201cecdb06db2a3dab7b9908b6/jax/_src/lax/linalg.py#L657).
The [TPU path]((https://github.com/google/jax/blob/f461c4ef0c0c73201cecdb06db2a3dab7b9908b6/jax/_src/lax/linalg.py#L685)) takes a different route that falls back to the non-MLIR XLA backend and then imports into MLIR. It still hits a custom call in `eigh_jacobi` which it uses to implement `eigh` ([here](https://github.com/google/jax/blob/f461c4ef0c0c73201cecdb06db2a3dab7b9908b6/jax/_src/lax/linalg.py#L601)). Here is the IR it produces if of interest ([IR](https://gist.github.com/silvasean/d455afd26eec00910ee36cc354d7cd7c))

Used by AlphaFold [here](https://github.com/deepmind/alphafold/blob/2e6a78f08b3a31945ad0fad77f7cf4344ae73d13/alphafold/model/quat_affine.py#L112)

## Steps to Reproduce the Problem

1. Run program above

## Specifications

- Version: 2ee66e0c0a8c4044281ca43015703eefff4ea958
- Platform: DISTRIB_DESCRIPTION="Ubuntu 22.04.2 LTS"

**Comment by silvasean** [comment]
@hawkinsp do you know of any general way I can "fallback to another backend / host computation" or something for cases like this? It seems like the only way to deal with this issue (which is blocking finding further issues) is to either

1. "solve it properly" (which could take a while)
2. open-code a Python implementation -- do you know of any library of reference implementations I can quickly port to use here?

**Comment by hawkinsp** [comment]
Well, in general, you're going to have the best time if you implement calling out to other libraries ("mhlo.custom_call"). That's what JAX's CPU and GPU paths do, as you note. While it might be possible to implement a performant `eigh` on CPU, it's not a trivial project.

That said if you're dead set on not doing that (seriously, reconsider), the TPU implementation in JAX can actually run without calling into any other `eigh` custom calls. Set `termination_size` to 1 here:
https://github.com/google/jax/blob/a9e48af260dd6d583c29fee87c472164dec077b9/jax/_src/lax/eigh.py#L421
and replace the base case here with the (trivial) eigendecomposition of a 1x1 matrix:
https://github.com/google/jax/blob/a9e48af260dd6d583c29fee87c472164dec077b9/jax/_src/lax/eigh.py#L341
This forces the implementation to use a divide-and-conquer algorithm down to size 1x1, which is comically small, but should work.

The catch here, is that you likely need at least two other primitives implemented: `cholesky` and `qr` decompositions. And I'm not entirely sure how you would implement them without custom call support.

Another option might be to use a Jacobi eigh exclusively. They work well at small sizes. You can find the implementation that backs the `eigh_jacobi` custom call on TPU here as a (classic) HLO to (classic) HLO rewrite:
https://github.com/openxla/xla/blob/main/xla/service/eigh_expander.cc
(The TPU implementation is optimized further over that implementation, but implements the same algorithm.)


But really, do this as a library call? LAPACK has a better implementation than you're going to come up with on CPU in reasonable time; it's already available in JAX, and we'd just need to make it available to the IREE plugin also.

**Comment by silvasean** [comment]
Thanks! I totally agree that library calls are the right solution here, and I'm trying to bring this bug to the attention of the right folks to get the library call mechanism implemented.

However, my current workstream is to prefetch issues and flush out unknowns/priorities throughout the stack, so quick workarounds are something I'm interested in as well (since the proper custom_call mechanism is at least 1-2 months away). Currently I'm swapping out with:

```
def _eigh_incorrect_impl(x, *, lower, sort_eigenvalues):
  # Totally incorrect -- just return "ones" of the correct shape
  *leading_dims, m, n = x.shape
  assert m == n, (m, n)
  return jnp.ones(x.shape, dtype=x.dtype), jnp.ones(tuple(leading_dims) + (m,), dtype=x.dtype)
```

in order to make further progress. I tried your suggestions for `termination_size` and indeed we hit a QR `custom_call` as you predicted.

Even if it is super slow, does JAX have any way for me to say "fall back to another device for this one op" or something like that?

**Comment by hawkinsp** [comment]
Not really. The way JAX has to call out to something else in the middle of a computation *is* `mhlo.custom_call`.

You could break the computation up into pieces and run that one `eigh` op on the current CPU path. That would work. It may be difficult to do. To do that, just place its inputs on `jax.devices("cpu")[0]` with `jax.device_put`.

**Comment by stellaraccident** [comment]
Sg. Let's just sort this into "needs a custom_call bucket". Likely assign it to Murali.

**Comment by silvasean** [comment]
cc @vmurali
